relay-server/src/processing/spans/mod.rs

```diff
@@ -365,33 +469,195 @@ pub struct Indexed;
 }
 
-impl CountRateLimited for Managed<ExpandedSpans<TotalAndIndexed>> {
+impl RateLimited for Managed<ExpandedSpans<TotalAndIndexed>> {
     type Error = Error;
+
+    async fn enforce<T>(
+        &mut self,
+        mut rate_limiter: T,
+        _: Context<'_>,
+    ) -> std::result::Result<(), Rejected<Self::Error>>
+    where
+        T: processing::RateLimiter,
+    {
+        let scoping = self.scoping();
+
+        let ExpandedSpansQuantities {
+            span,
+            span_indexed,
+            attachment,
+            attachment_item,
+        } = self.quantities_helper();
+
+        if span > 0 {
+            let limits = rate_limiter
+                .try_consume(scoping.item(DataCategory::Span), span)
+                .await;
+            if !limits.is_empty() {
+                // If there is a span quota reject all the spans and the associated attachments.
+                return Err(self.reject_err(Error::from(limits)));
+            }
+        }
+
+        if span_indexed > 0 {
+            let limits = rate_limiter
+                .try_consume(scoping.item(DataCategory::SpanIndexed), span_indexed)
+                .await;
+            if !limits.is_empty() {
+                // If there is a span quota reject all the spans and the associated attachments.
+                return Err(self.reject_err(Error::from(limits)));
+            }
+        }
+
+        if attachment > 0 {
+            let limits = rate_limiter
+                .try_consume(scoping.item(DataCategory::Attachment), attachment)
+                .await;
+
+            if !limits.is_empty() {
+                self.modify(|this, record_keeper| {
+                    // Reject both stand_alone and associated attachments.
+                    let mut all_attachments = std::mem::take(&mut this.stand_alone_attachments);
+                    all_attachments.extend(
+                        this.spans
+                            .iter_mut()
+                            .filter_map(|span| span.attachment.take()),
+                    );
+
+                    record_keeper.reject_err(Error::from(limits), all_attachments);
+                });
+            }
+        }
+
+        if attachment_item > 0 {
+            let limits = rate_limiter
+                .try_consume(scoping.item(DataCategory::AttachmentItem), attachment_item)
+                .await;
+
+            if !limits.is_empty() {
+                self.modify(|this, record_keeper| {
+                    // Reject both stand_alone and associated attachments.
+                    let mut all_attachments = std::mem::take(&mut this.stand_alone_attachments);
+                    all_attachments.extend(
+                        this.spans
+                            .iter_mut()
+                            .filter_map(|span| span.attachment.take()),
+                    );
+
+                    record_keeper.reject_err(Error::from(limits), all_attachments);
+                });
+            }
+        }
```

Comment:
nit: This could be a helper function to deduplicate the code with the `attachments` check.